hyperapp.d.ts

```diff
@@ -51,58 +56,126 @@ export function h<Props>(
  */
-export type ActionType<State, Actions> = (
-  data?: any
-) =>
-  | ((state: State, actions: Actions) => ActionResult<State>)
-  | ActionResult<State>
+type ActionProxy<
+  TState extends StateType,
+  TActions extends UnwiredActions<TState, TActions>
+> = { [P in keyof TActions]: WiredAction<TState, any> }
 
-/** The interface for actions implementations.
+/** 
+ * Type to describe anything that's not a function or object
+ */
+export type Primitive = string | number | boolean | null | undefined | any[]
+
+/** 
+ * The result of an action.
  *
  * @memberOf [App]
  */
-export type ActionsType<State, Actions> = {
-  [P in keyof Actions]:
-    | ActionType<State, Actions>
-    | ActionsType<any, Actions[P]>
-}
+export type ActionResult<TState extends Primitive | NestedMap<Primitive>> =
+  | Partial<TState>
+  | Promise<any>
+  | null
+  | void
 
-/** The view function describes the application UI as a tree of VNodes.
+/** 
+ * The view function describes the application UI as a tree of VNodes.
  * @returns A VNode tree.
  * @memberOf [App]
  */
-export interface View<State, Actions> {
-  (state: State, actions: Actions): VNode<object>
+export interface View<
+  TState extends StateType,
+  TActions extends WiredActions<TState, TActions>
+> {
+  (state: TState, actions: TActions): VNode<any>
 }
 
-/** The app() call creates and renders a new application.
+/** 
+ * The app() call creates and renders a new application.
  *
  * @param state The state object.
  * @param actions The actions object implementation.
  * @param view The view function.
  * @param container The DOM element where the app will be rendered to.
  * @returns The actions wired to the application.
- * @memberOf [App]
  */
-export function app<State, Actions>(
-  state: State,
-  actions: ActionsType<State, Actions>,
-  view: View<State, Actions>,
-  container: Element | null
-): Actions
+export declare const app: App
 
-/** @namespace [JSX] */
+/**
+ * The interface for app(). Use this for implementing Higher Order App's
+ */
+export interface App {
+  <TState extends StateType, TActions extends UnwiredActions<TState, TActions>>(
+    state: TState,
+    actions: TActions,
+    view: View<TState, ActionProxy<TState, TActions>>,
+    container: HTMLElement | null
+  ): ActionProxy<TState, TActions>
+}
 
+/** 
+ * The type of action that's been returned from app()
+ * 
+ * This action is also passed as second parameter to the function that's 
+ * returned from unwired actions.
+ */
+type WiredAction<
+  TState extends Primitive | NestedMap<Primitive>,
+  TPayload extends Primitive | NestedMap<Primitive> = any
+> = (data?: TPayload) => TState
+
+/** 
+ * The action that's being passed as second param in an object to app(). 
+ * Use this when implementing your actions.
+ */
+export type UnwiredAction<
+  TState extends StateType,
+  TActions extends UnwiredActions<TState, TActions>,
+  TPayload extends Primitive | NestedMap<Primitive> = any
+> = (
+  data?: TPayload
```

Comment:
What is TPayload?

Reply:
The generic to type the payload

Reply:
What is the payload?

Reply:
The payload passed to an action when calling from the view or inside an action.

Reply:
Redux also calls it payload and it's called payload in the Flux architecture. I believe it's wise to keep it the same as the rest of the community calls it.

Reply:
@Alber70g Ah, the data? 

```js
actions.foo(payload)
```
?

Reply:
Yes the Data, but I think payload makes sense since it is used as a name in Redux too.

Reply:
Yes that's correct.

Reply:
I agree with this! Thanks for explaining.